File: apps/api/src/rag/vectorStore.ts
import { ChromaClient, type Collection, IncludeEnum, type Metadata } from 'chromadb';
import { env } from '../config/env';
import { logger } from '../config/logger';

/** Topics / collections yang diinisialisasi saat startup. */
export const RAG_TOPICS = [
  'harga',
  'layanan',
  'gaya-rambut',
  'tips-perawatan',
  'booking-info',
] as const;

export type RagTopic = (typeof RAG_TOPICS)[number];

export interface VectorRecord {
  id: string;
  text: string;
  embedding: number[];
  metadata: Record<string, string | number>;
}

export interface VectorHit {
  id: string;
  text: string;
  distance: number;
  metadata: Record<string, string | number>;
}

/**
 * Wrapper persistent ChromaDB.
 * - Koneksi: HTTP ke server ChromaDB (CHROMA_URL). Persist dir server dikonfigurasi
 *   saat menjalankan server (mis. `chroma run --path ./chroma_db`).
 * - Collection satu per topic (harga, layanan, gaya-rambut, tips-perawatan, booking-info).
 */
export class VectorStore {
  private client: ChromaClient;
  private collections = new Map<RagTopic, Collection>();

  constructor(url: string = env.CHROMA_URL) {
    this.client = new ChromaClient({ path: url });
  }

  async init(): Promise<void> {
    for (const topic of RAG_TOPICS) {
      const col = await this.client.getOrCreateCollection({
        name: `rag-salon-${topic}`,
      });
      this.collections.set(topic, col);
    }
    logger.info({ topics: Array.from(this.collections.keys()) }, 'VectorStore: collection siap');
  }

  private async collection(topic: RagTopic): Promise<Collection> {
    let col = this.collections.get(topic);
    if (!col) {
      col = await this.client.getOrCreateCollection({ name: `rag-salon-${topic}` });
      this.collections.set(topic, col);
    }
    return col;
  }

  async upsert(topic: RagTopic, records: VectorRecord[]): Promise<void> {
    if (records.length === 0) return;
    const col = await this.collection(topic);
    await col.upsert({
      ids: records.map((r) => r.id),
      embeddings: records.map((r) => r.embedding),
      documents: records.map((r) => r.text),
      metadatas: records.map((r) => r.metadata as Metadata),
    });
    logger.debug({ topic, count: records.length }, 'VectorStore.upsert');
  }

  /** Query embedding terhadap satu topic. */
  async query(topic: RagTopic, embedding: number[], topK: number): Promise<VectorHit[]> {
    const col = await this.collection(topic);
    const res = await col.query({
      queryEmbeddings: embedding,
      nResults: topK,
      include: [IncludeEnum.Documents, IncludeEnum.Metadatas, IncludeEnum.Distances],
    });
    const docs = res.documents?.[0] ?? [];
    const dist = res.distances?.[0] ?? [];
    const metas = res.metadatas?.[0] ?? [];
    return docs.map((doc, i) => ({
      id: (res.ids?.[0]?.[i] as string) ?? String(i),
      text: doc ?? '',
      distance: dist[i] ?? 0,
      metadata: (metas[i] as Record<string, string | number> | null) ?? {},
    }));
  }

  /** Ambil seluruh id yang tersimpan (untuk diff idempotent). */
  async existingIds(topic: RagTopic): Promise<string[]> {
    const col = await this.collection(topic);
    const res = await col.get({ include: [] });
    return res.ids as string[];
  }

  /** Hapus record spesifik berdasarkan id. */
  async deleteByIds(topic: RagTopic, ids: string[]): Promise<void> {
    if (ids.length === 0) return;
    const col = await this.collection(topic);
    await col.delete({ ids });
  }

  /** Hapus seluruh records (utility, mis. re-ingest bersih). */
  async clear(topic: RagTopic): Promise<void> {
    const existing = await this.existingIds(topic);
    if (existing.length > 0) {
      const col = await this.collection(topic);
      await col.delete({ ids: existing });
    }
  }
}

export const vectorStore = new VectorStore();

/** Panggil saat bootstrap (app.ts) — koneksi & prep collection; gagal non-fatal. */
export function initVectorStore(): void {
  void vectorStore.init().catch((err) => {
    logger.warn(
      { err: (err as Error).message },
      'ChromaDB tidak tersedia; RAG menunggu ingest/manual init',
    );
  });
}

File: apps/ai/app/rag/vector_store.py
"""ChromaDB vector store — native Python (dual-mode persistent | http).

Port of `apps/api/src/rag/vectorStore.ts` to the Python chromadb client.
Collection one per topic (harga, layanan, gaya-rambut, tips-perawatan, booking-info).
"""

from __future__ import annotations

import logging
from typing import Optional

import chromadb
from chromadb.api.models.Collection import Collection

from app.config import settings

logger = logging.getLogger(__name__)

# Koleksi KB publik (chat pelanggan) — TANPA katalog admin.
RAG_TOPICS = [
    "harga",
    "layanan",
    "gaya-rambut",
    "tips-perawatan",
    "booking-info",
]

# Koleksi untuk agent admin (termasuk katalog hasil serialize DB).
AGENT_TOPICS = [*RAG_TOPICS, "katalog"]


def collection_name(topic: str) -> str:
    return f"rag-salon-{topic}"


class VectorStore:
    def __init__(self) -> None:
        self._client: Optional[chromadb.ClientAPI] = None
        self._collections: dict[str, Collection] = {}

    def _get_client(self):
        if self._client is None:
            if settings.chroma_mode == "http":
                self._client = chromadb.HttpClient(host=settings.chroma_url, port=8000)
            else:
                settings.chroma_persist_dir.mkdir(parents=True, exist_ok=True)
                self._client = chromadb.PersistentClient(path=str(settings.chroma_persist_dir))
            logger.info("ChromaDB client siap (mode=%s)", settings.chroma_mode)
        return self._client

    def _collection(self, topic: str) -> Collection:
        col = self._collections.get(topic)
        if col is None:
            col = self._get_client().get_or_create_collection(name=collection_name(topic))
            self._collections[topic] = col
        return col

    def upsert(
        self,
        topic: str,
        ids: list[str],
        texts: list[str],
        embeddings: list[list[float]],
        metadatas: Optional[list[dict]] = None,
    ) -> None:
        col = self._collection(topic)
        mds = metadatas or [{} for _ in ids]
        col.upsert(ids=ids, documents=texts, embeddings=embeddings, metadatas=mds)
        logger.debug("upsert topic=%s n=%d", topic, len(ids))

    async def query(
        self,
        topic: str,
        embedding: list[float],
        top_k: int,
    ) -> list[dict]:
        """Return hits: [{"id", "text", "distance", "metadata"}, ...]."""
        col = self._collection(topic)
        res = col.query(query_embeddings=[embedding], n_results=top_k, include=["documents", "metadatas", "distances"])
        ids = res.get("ids", [[]])[0]
        docs = res.get("documents", [[]])[0]
        dists = res.get("distances", [[]])[0]
        metas = res.get("metadatas", [[]])[0]
        out: list[dict] = []
        for i, doc_id in enumerate(ids or []):
            out.append(
                {
                    "id": str(doc_id),
                    "text": docs[i] or "" if i < len(docs) else "",
                    "distance": float(dists[i]) if i < len(dists) else 0.0,
                    "metadata": metas[i] or {} if i < len(metas) else {},
                }
            )
        return out

    def existing_ids(self, topic: str) -> list[str]:
        col = self._collection(topic)
        return col.get(include=[])["ids"]

    def delete_by_ids(self, topic: str, ids: list[str]) -> None:
        if not ids:
            return
        col = self._collection(topic)
        col.delete(ids=ids)

    def clear(self, topic: str) -> None:
        ids = self.existing_ids(topic)
        if ids:
            self.delete_by_ids(topic, ids)


vector_store = VectorStore()